### PR TITLE
Bugfix/broken distinct

### PR DESCRIPTION
**Pull request by StanJav**
EqCompare shouldn't use the obj.GetHashCode(). The Equals is used only if the GetHashCode of two objects is same. Usually a list of objects doesn't include duplicate hashes but that doesn't mean that a specific field within the object isn't same.  A test added for this.

One more usefull distinct method added so you can rewrite this:
`    lst.Distinct((a, b) => a.Item1 == b.Item1)`
into this:
`    lst.Distinct(o => o.Item1)`

### CODE REVIEW

File: LanguageExt.Core/List.cs [hunk 1]
```diff
@@ -777,6 +777,16 @@ public static Unit iter<T>(IEnumerable<T> list, Action<int, T> action)
...
         /// <summary>
+        /// Return a new enumerable with all duplicate values removed
+        /// </summary>
+        /// <typeparam name="T">Enumerable item type</typeparam>
+        /// <param name="list">Enumerable</param>
+        /// <returns>A new enumerable with all duplicate values removed</returns>
+        [Pure]
+        public static IEnumerable<T> distinct<T, K>(IEnumerable<T> list, Func<T, K> keySelector, Option<Func<K, K, bool>> compare = default(Option<Func<K, K, bool>>)) =>
+             list.Distinct(new EqCompare<T>((a, b) => compare.IfNone(EqualityComparer<K>.Default.Equals)(keySelector(a), keySelector(b)), a => keySelector(a).GetHashCode()));
+
```
Comment (louthy) [inline LanguageExt.Core/List.cs]: `a => keySelector(a).GetHashCode()`  no null test, use: `a => keySelector(a)?.GetHashCode() ?? 0`

File: LanguageExt.Core/List.cs [hunk 2]
```diff
@@ -1064,12 +1074,19 @@ public static IEnumerable<IEnumerable<T>> tails<T>(IEnumerable<T> self)
     class EqCompare<T> : IEqualityComparer<T>
     {
         readonly Func<T, T, bool> compare;
+        readonly Option<Func<T, int>> getHashCode = None;
```
Comment (louthy) [inline LanguageExt.Core/List.cs]: Renamed this to `hashCode` please, `getHashCode` feels nasty.

File: LanguageExt.Core/List.cs [hunk 3]
```diff
@@ -1064,12 +1074,19 @@ public static IEnumerable<IEnumerable<T>> tails<T>(IEnumerable<T> self)
...
 
         public EqCompare(Func<T, T, bool> compare)
         {
             this.compare = compare;
         }
 
+        public EqCompare(Func<T, T, bool> compare, Func<T, int> getHashCode)
```
Comment (louthy) [inline LanguageExt.Core/List.cs]: `getHashCode` -> `hashCode`

File: LanguageExt.Core/Queue.cs [hunk 4]
```diff
@@ -266,6 +270,10 @@ public static class QueueExtensions
         [Pure]
         public static IEnumerable<T> Distinct<T>(Que<T> queue, Func<T, T, bool> compare) =>
             LanguageExt.List.distinct(queue, compare);
+        
+        [Pure]
+        public static IEnumerable<T> Distinct<T, K>(Que<T> queue, Func<T, K> keySelector, Option<Func<K, K, bool>> compare = default(Option<Func<K, K, bool>>)) =>
```
Comment (louthy) [inline LanguageExt.Core/Queue.cs]: Is this needed?  `Que<T>` is already an `IEnumerable`, and therefore the `Distinct` extension method for `IEnumerable` will do the same job.  If it is needed, then it should be `this Que<T>` as the first argument.

File: LanguageExt.Core/Stack.cs [hunk 5]
```diff
@@ -840,6 +850,16 @@ public static class StackExtensions
...
 
     /// <summary>
+    /// Return an enumerable with all duplicate values removed
+    /// </summary>
+    /// <typeparam name="T">Stack item type</typeparam>
+    /// <param name="stack">Stack</param>
+    /// <returns>An enumerable with all duplicate values removed</returns>
+    [Pure]
+    public static IEnumerable<T> Distinct<T, K>(Stck<T> stack, Func<T, K> keySelector, Option<Func<K, K, bool>> compare = default(Option<Func<K, K, bool>>)) =>
+        LanguageExt.List.distinct(stack, keySelector, compare);
```
Comment (louthy) [inline LanguageExt.Core/Stack.cs]: Is this needed?  `Stck<T>` is already an `IEnumerable`, and therefore the `Distinct` extension method for `IEnumerable` will do the same job.  If it is needed, then it should be `this Stck<T>` as the first argument.

File: LanguageExt.Core/List.cs (hunk 1 above)
Comment (StanJav) [inline LanguageExt.Core/List.cs]: Done.

File: LanguageExt.Core/List.cs (hunk 2 above)
Comment (StanJav) [inline LanguageExt.Core/List.cs]: Done.

File: LanguageExt.Core/List.cs (hunk 3 above)
Comment (StanJav) [inline LanguageExt.Core/List.cs]: Done.

File: LanguageExt.Core/Queue.cs (hunk 4 above)
Comment (StanJav) [inline LanguageExt.Core/Queue.cs]: Copy & paste  issue. Done also for the existing Distinct and Edists.

File: LanguageExt.Core/Stack.cs (hunk 5 above)
Comment (StanJav) [inline LanguageExt.Core/Stack.cs]: Copy & paste  issue. Done also for the existing Exists.

**Comment by louthy** [comment]
@StanJav  Cheers.  How quickly do you need it?

**Comment by StanJav** [comment]
>How quickly do you need it?

@louthy This year please.

**Comment by StanJav** [comment]
Hi @louthy , could you please build a new version of lang-ext?

**Pull request merged**